### PR TITLE
Allow 'me' literal to represent the current user in url

### PR DESCRIPTION
**Pull request by raymondfeng**
/to @ritch @bajtos 

This PR adds the ability to use `me` or a custom literal to represent the current user in req url. For example,

GET /api/users/me

The token middleware will replace `me` with the current user id for downstream handlers.

It adds an option to loopback.token middleware as `currentUserLiteral`.

### CODE REVIEW

**Comment by ritch** [comment]
We need to make it clear that you can use `req.originalUrl` to differentiate between `/users/me` and `/users/7`.

**Comment by ritch** [comment]
We need to add some examples using this to the docs.

**Comment by ritch** [comment]
We also need to update the `loopback.token()` reference to include this option.

**Comment by raymondfeng** [comment]
@bajtos Any comments before I merge?

File: server/middleware/token.js [hunk 1]
```diff
@@ -11,6 +11,18 @@ var assert = require('assert');
...
 module.exports = token;
 
+/*
+ * Rewrite the url to replace current user literal with the logged in user id
+ */
+function rewriteUserLiteral(req, currentUserLiteral) {
+  if (req.accessToken && req.accessToken.userId && currentUserLiteral) {
+    // Replace /me/ with /current-user-id/
+    req.url = req.url.replace(
+      new RegExp('/' + currentUserLiteral + '(/|$|\\?)', 'g'),
```
Comment (bajtos) [inline server/middleware/token.js]: This assumes "currentUserLiteral" does not contain any special regexp characters like ".". This requirement should be captured in the docs and API docs.

Comment (bajtos) [inline server/middleware/token.js]: I am also wondering why are you replacing all occurrences of "/me" via the "g" flag?

File: server/middleware/token.js [hunk 2]
```diff
@@ -39,19 +51,33 @@ module.exports = token;
...
  */
 
 function token(options) {
   options = options || {};
   var TokenModel = options.model || loopback.AccessToken;
-  assert(TokenModel, 'loopback.token() middleware requires a AccessToken model');
+  if (typeof TokenModel === 'string') {
+    // Make it possible to configure the model in middleware.json
+    TokenModel = loopback.getModel(TokenModel);
+  }
```
Comment (bajtos) [inline server/middleware/token.js]: Unrelated change, please move it to a standalone commit (or even a PR).

File: server/middleware/token.js [hunk 3]
```diff
@@ -39,19 +51,33 @@ module.exports = token;
...
+  var currentUserLiteral = options.currentUserLiteral;
+  if (currentUserLiteral && (typeof currentUserLiteral !== 'string')) {
```
Comment (bajtos) [inline server/middleware/token.js]: You should report a warning (at least via `debug`) if currentLiteralString is defined and not a string. Otherwise users can be confused why their setting was ignored.

Should we support `currentLiteralString: false` to disable this feature?

File: test/access-token.test.js [hunk 4]
```diff
@@ -108,6 +108,48 @@ describe('loopback.token(options)', function() {
...
+        .get('/users/me')
+        .set('authorization', id)
+        .end(function(err, res) {
+          assert(!err);
+          assert.deepEqual(res.body, {userId: '123'});
+          done();
+        });
+    });
+
+  it('should rewrite url for the current user literal at the end',
```
Comment (bajtos) [inline test/access-token.test.js]: This is the same name as on L111, please make it unique and descriptive.

File: test/access-token.test.js [hunk 5]
```diff
@@ -108,6 +108,48 @@ describe('loopback.token(options)', function() {
...
+    function(done) {
+      var app = createTestApp(this.token, done);
+      var id = this.token.id;
+      request(app)
+        .get('/users/me')
+        .set('authorization', id)
+        .end(function(err, res) {
+          assert(!err);
+          assert.deepEqual(res.body, {userId: '123'});
```
Comment (bajtos) [inline test/access-token.test.js]: You should not use magic values like `123` that are specified deep in test helpers. You should be able to read the user id from `token.userId`, right?

Please fix this issue in all other tests too.

File: test/access-token.test.js [hunk 6]
```diff
@@ -284,7 +326,7 @@ describe('app.enableAuth()', function() {
 
 function createTestingToken(done) {
   var test = this;
-  Token.create({}, function(err, token) {
+  Token.create({userId: '123'}, function(err, token) {
```
Comment (bajtos) [inline test/access-token.test.js]: See my previous comment, this should not be needed.

**Comment by bajtos** [comment]
A couple more things to consider.

1) Is the middleware the right place for this feature? What if the app is using JSON-RPC or WebSockets? Shouldn't this be implemented by `User.findById` instead?

2) Is it a good idea to shoehorn this into token middleware? Shouldn't this be a standalone middleware, e.g. `loopback.currentUserUrl`?

File: server/middleware/token.js (hunk 3 above)
Comment (raymondfeng) [inline server/middleware/token.js]: By default, this feature is NOT enabled as `currentUserLiteral === undefined`. Setting `currentUserLiteral` to a `falsey` value will disable it too.

**Comment by bajtos** [comment]
> We need to make it clear that you can use req.originalUrl to differentiate between /users/me and /users/7.

IIRC, `originalUrl` is a full path including any prefixes stripped by outside apps, i.e. if you mount REST at "/api", the rest adapter will see `url: /users/1234` and `originalUrl: /api/users/1234`. In which case it may be worth to store the "me" url in a new property. I may be wrong though, feel free to ignore this comment.

File: server/middleware/token.js (hunk 3 above)
Comment (bajtos) [inline server/middleware/token.js]: > By default, this feature is NOT enabled as currentUserLiteral === undefined. Setting currentUserLiteral to a falsey value will disable it too.

I see, makes sense.

I still think we should let the user know when we decide to ignore his config option, e.g. because it's an object or an array.

**Comment by raymondfeng** [comment]
The `originalUrl` is the full path and it won't be rewritten.

**Comment by bajtos** [comment]
> The originalUrl is the full path and it won't be rewritten.

This is my point: If my assumptions are correct, then there is no way for the rest adapter to get the unmodified local url, i.e. "/users/me". That's why I am proposing to save this URL in a new property.

File: test/access-token.test.js (hunk 6 above)
Comment (raymondfeng) [inline test/access-token.test.js]: This is needed to make sure there is a userId property in the token instance.

Comment (bajtos) [inline test/access-token.test.js]: Fair enough, just don't use the magic constant `123` in test assertions please.

**Comment by raymondfeng** [comment]
> 1) Is the middleware the right place for this feature? What if the app is using JSON-RPC or WebSockets? Shouldn't this be implemented by User.findById instead?

The currentUserLiteral rewrite is only supported for the token-based authentication. It doesn't make sense to implement it in `User.findById` because the `me` context is not a concern of `User.findById`.

> 2) Is it a good idea to shoehorn this into token middleware? Shouldn't this be a standalone middleware, e.g. loopback.currentUserUrl?

In the longer term, I'm open to refactor it into a `loopback.urlrewrite` middleware which will support more rules beyond `currentUserLiteral`. Any rules that reference the `currentUserLiteral` can only be activated in a phase after `auth` though. For example, 
```json
"intial": {
    "loopback#urlrewrite": {
      "params": {
        "rules": [...]
      }
    }
  },
"auth:after": {
    "loopback#urlrewrite": {
      "params": {
        "currentUserLiteral": "me",
        "rules": [...]
      }
    }
  }
```

**Comment by bajtos** [comment]
> The currentUserLiteral rewrite is only supported for the token-based authentication. It doesn't make sense to implement it in User.findById because the me context is not a concern of User.findById.

How about a "beforeRemote" hook then? It has access to the parsed value of "id" parameter and also access to the request context (to get the current user id).

The implementation proposed in this pull request does not discriminate between different places where the "me" string can occur in the URL, thus it's possible it will modify non-user paths too.

I don't like the implementation proposed here from the reasons above. However, if you believe this is the best solution possible, then I can live with it and you can land this patch as it is now.

**Comment by bajtos** [comment]
> you can land this patch as it is now.

After you fix the failing build please.

**Comment by raymondfeng** [comment]
The build error should be fixed. It was due to a style issue.

**Comment by raymondfeng** [comment]
> How about a "beforeRemote" hook then? It has access to the parsed value of "id" parameter and also access to the request context (to get the current user id).
The implementation proposed in this pull request does not discriminate between different places where the "me" string can occur in the URL, thus it's possible it will modify non-user paths too.

It depends on the scope of url rewrite. If we just want to deal with user model (only /api/users/:id/...), it can be implemented as a beforeRemote hook. I see cases that `me` can be used beyond the user model apis, including the api gateway with oAuth 2. We also need to understand if the url rewrite should happen before other middleware that rely on the userId.

File: server/middleware/token.js (hunk 2 above)
Comment (bajtos) [inline server/middleware/token.js]: @raymondfeng how about this part? ^^^

**Comment by bajtos** [comment]
> It depends on the scope of url rewrite. If we just want to deal with user model (only /api/users/:id/...), it can be implemented as a beforeRemote hook. I see cases that me can be used beyond the user model apis, including the api gateway with oAuth 2. We also need to understand if the url rewrite should happen before other middleware that rely on the userId.

Fair enough. My primary goal was to ensure that we have considered all options and picked the best one.

I don't like the unrelated change squashed into this patch, but that's a detail I can live with.

LGTM.

**Comment by james-relyea** [comment]
Currently, if you use an expired/invalid access token when accessing relations of the user model, using the 'me' literal causes a 404 error instead of 401. Is this intentional? To reproduce:

GET /api/users/<user-id>/<some-relation>?access_token=<valid_access_token>, returns 200
GET /api/users/me/<some-relation>?access_token=<valid_access_token>, returns 200

GET /api/users/<user-id>/<some-relation>?access_token=<invalid_access_token>, returns 401
GET /api/users/me/<some-relation>?access_token=<invalid_access_token>, returns 404

Should this be a 404 if the path exists for authenticated users? I would think that if using the 'me' literal requires an authenticated user, it should return a 401 if the authentication fails.

**Comment by james-relyea** [comment]
@raymondfeng ^^
